pilot/pkg/bootstrap/server.go

```diff
@@ -246,25 +249,83 @@ func NewServer(args *PilotArgs, initFuncs ...func(*Server)) (*Server, error) {
 		return nil, err
 	}
 
-	if err := s.initControllers(args); err != nil {
+	// Parse and validate Istiod Address.
+	istiodHost, _, err := e.GetDiscoveryAddress()
+	if err != nil {
 		return nil, err
 	}
 
-	s.XDSServer.InitGenerators(e, args.Namespace)
-
-	// Initialize workloadTrustBundle after CA has been initialized
-	if err := s.initWorkloadTrustBundle(args); err != nil {
+	// Create Istiod certs and setup watches.
+	if err := s.initIstiodCerts(args, string(istiodHost)); err != nil {
 		return nil, err
 	}
 
-	// Parse and validate Istiod Address.
-	istiodHost, _, err := e.GetDiscoveryAddress()
-	if err != nil {
+	// federation support must be initialized before config and service controllers
+	if features.EnableFederation {
+		if s.kubeClient == nil {
+			log.Errorf("could not initialize federation discovery server: kubeClient is nil")
+		} else {
```

Comment:
I think we could encapsulate this initialization to an unexported function to hide all its details and be consistent with other init functions.